Account for LLM provider returning a negative tool call index

Sherlock-Holo opened the pull request:
some inference framework may return `-1` `index` when returning tool call response, and some other one will not return the `index` field. make the index type as `Option<isize>` can handle these problems

Sherlock-Holo committed:
fix tool call

some inference framework may return -1 index when returning tool call
response, and some will not return the index field. make the index type
as Option<isize> can handle these problems

Sherlock-Holo commented:
@cla-bot check

benbrandt commented on crates/language_models/src/provider/lmstudio.rs:
```diff
@@ -531,7 +531,11 @@ impl LmStudioEventMapper {
-                let entry = self.tool_calls_by_index.entry(tool_call.index).or_default();
+                let mut index = tool_call.index.unwrap_or_default();
+                if index < 0 {
+                    index = 0;
```
Why are we defaulting to 0 here? 
I would expect -1 to either mean it doesn't match or it is the last one, not the first. If you could provide the docs where this is the case that would help.

But looking at the vllm example you posted, it seems this was just a helper to force logging after certain runs of the loop (an odd way to do it, but I don't think the intention was to actually have a -1 index)

It does feel like it might be better to report the issue to z.ai, we've had a fair number of spec incompatibilities for that provider it seems.

Sherlock-Holo committed:
fix too call

self-hosted vllm may not return index in tool call response, also, the tool call id
and function name may empty. I don't know why, but this change can make
zed support self-hosted vllm qwen3-coder model
